pkg/eth/retriever.go

```diff
@@ -0,0 +1,132 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	"database/sql"
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/ipfs-blockchain-watcher/pkg/postgres"
+)
+
+// Retriever interface for substituting mocks in tests
+type Retriever interface {
+	RetrieveFirstBlockNumber() (int64, error)
+	RetrieveLastBlockNumber() (int64, error)
+	RetrieveGapsInData(validationLevel int) ([]DBGap, error)
+}
+
+// GapRetriever type for Ethereum
+type GapRetriever struct {
+	db *postgres.DB
+}
+
+// NewGapRetriever returns a pointer to a new GapRetriever
+func NewGapRetriever(db *postgres.DB) *GapRetriever {
+	return &GapRetriever{
+		db: db,
+	}
+}
+
+// RetrieveFirstBlockNumber is used to retrieve the first block number in the db
+func (ecr *GapRetriever) RetrieveFirstBlockNumber() (int64, error) {
+	var blockNumber int64
+	err := ecr.db.Get(&blockNumber, "SELECT block_number FROM eth.header_cids ORDER BY block_number ASC LIMIT 1")
+	return blockNumber, err
+}
+
+// RetrieveLastBlockNumber is used to retrieve the latest block number in the db
+func (ecr *GapRetriever) RetrieveLastBlockNumber() (int64, error) {
+	var blockNumber int64
+	err := ecr.db.Get(&blockNumber, "SELECT block_number FROM eth.header_cids ORDER BY block_number DESC LIMIT 1 ")
+	return blockNumber, err
+}
+
+// DBGap type for querying for gaps in db
+type DBGap struct {
+	Start uint64 `db:"start"`
+	Stop  uint64 `db:"stop"`
+}
+
+// RetrieveGapsInData is used to find the the block numbers at which we are missing data in the db
+// it finds the union of heights where no data exists and where the times_validated is lower than the validation level
+func (ecr *GapRetriever) RetrieveGapsInData(validationLevel int) ([]DBGap, error) {
+	log.Info("searching for gaps in the eth ipfs watcher database")
+	startingBlock, err := ecr.RetrieveFirstBlockNumber()
+	if err != nil {
+		return nil, fmt.Errorf("eth CIDRetriever RetrieveFirstBlockNumber error: %v", err)
+	}
+	var initialGap []DBGap
+	if startingBlock != 0 {
+		stop := uint64(startingBlock - 1)
+		log.Infof("found gap at the beginning of the eth sync from 0 to %d", stop)
```

Comment:
is it correct to put `if startingBlock != 0 ` and `found gap at the beginning` together? 
If startingBlock != 0 it means we are not at the beginning?

Reply:
If the earliest block we find in the db is *not* at height 0 then there is a gap between height 0 and the lowest block that *was* found